universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,206 @@
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager at " << &manager;
+    return manager;
+}
+
+
+unsigned int NamedValueRefManager::GetCheckSum() const {
+    unsigned int retval{0};
+    for (auto const& name_type_pair : m_value_refs)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    for (auto const& name_type_pair : m_value_refs_int)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    for (auto const& name_type_pair : m_value_refs_double)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    //CheckSums::CheckSumCombine(retval, m_value_refs.size()); // why also add size?
+    
+    DebugLogger() << "NamedValueRefManager checksum: " << retval;
+    return retval;
+}
+
+
+NamedValueRefManager::any_container_type  NamedValueRefManager::GetItems() const {
+    std::function<const any_entry_type(const entry_type&)> base_to_base          = [](const entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+    std::function<const any_entry_type(const double_entry_type&)> double_to_base = [](const double_entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+    std::function<const any_entry_type(const int_entry_type&)> int_to_base       = [](const int_entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+
+    // should use C++20 ranges (or C++14, C++17 boost Range-v3 library) to avoid copying
+    any_container_type aet;
+    std::transform(m_value_refs_double.begin(), m_value_refs_double.end(), std::inserter(aet, aet.end()), double_to_base);
+    std::transform(m_value_refs_int.begin(), m_value_refs_int.end(), std::inserter(aet, aet.end()), int_to_base);
+    std::transform(m_value_refs.begin(), m_value_refs.end(), std::inserter(aet, aet.end()), base_to_base);
+    for (auto kv : aet)
+        ErrorLogger() << "NamedValueRefManager::GetItems - " << kv.first << " " << kv.second.get().StringResult() << " := " << kv.second.get().Description() ;
+    return aet;
+}
+
+template <typename R, typename VR>
+void NamedValueRefManager::RegisterValueRefImpl(R& container, std::mutex& mutex, const std::string&& label, std::string&& valueref_name, std::unique_ptr<VR>&& vref) {
+    InfoLogger() << "Register " << label << " valueref for " << valueref_name << ": " << vref->Description();
+    if (container.count(valueref_name)>0) {
+        DebugLogger() << "Skip registration for already registered " << label << " valueref for " << valueref_name;
+        DebugLogger() << "Number of registered " << label << " ValueRefs: " << m_value_refs.size();
+        return;
+    }
+    const std::lock_guard<std::mutex> lock(mutex);
+    if (!(vref->RootCandidateInvariant() && vref->LocalCandidateInvariant() && vref->TargetInvariant() && vref->SourceInvariant()))
+            ErrorLogger() << "Currently only invariant value refs can be named. " << valueref_name;
+    container.emplace(valueref_name, std::move(vref));
```

Comment:
`std::move(valueref_name)`

Reply:
I thought because this s a key in a map it means it is const. cant move into a const, it would do a copy(?)
OTOH the emplace signature has only rvalue references and std::pair also seems to support that. so i move.

Reply:
you can't modify a key in a map, but you can move-construct one